### ISSUE TITLE
variable num_steps in custom rnn

### ISSUE THREAD

**Issue by rajarsheem** [NHDaly/jupyter-binder#1]
How can I use variable num_steps in the custom rnn so that I can feed  sequences of variable length at each iteration ?

**Comment by NHDaly** [comment]
I'm assuming you're talking about the Tensorflow custom rnn?

Note that this file was just for demonstrating and learning about RNNs. If you are trying to do any actual productionized machine learning, you should consider using something like Tensorflow's built in LSTM, which is like a more powerful recurrent neural network, I think. Their tutorial is here:
https://www.tensorflow.org/versions/r0.8/tutorials/recurrent/index.html

Huh, I hadn't thought about that. Why would you want to do this, can I ask? What do you think the benefits would be? (I'm genuinely curious, since I still don't know much about Neural Networks.) As long as you don't reset the hidden-state in-between batches, I think this should be okay? But it might also have some bad effects on the learning; I'm really not sure.

---

That said, it seems doable. My thoughts on how you would do this:

The variables can (maybe should) be initialized earlier than they are in my code right now. You can swap `In [6]` and `In [7]`, so that the variables are initialized as soon as they're created.

Then, the rest of the set-up could be parametrized with a variable `num_steps`. So every time you have a new batch, you could recreate the model based on this new `num_steps`, but keep the variables unmodified. Then you pass the old variables to the new model via `sess.run((train_step, hs[num_steps-1], cross_entropy), feed_dict=feed_dict)`.

So, I think it could look something like:

```
def RunBatch(input_chars, target_chars):  # length of two vectors must be equal.
    num_steps = len(input_chars)
    # Define model (Currently In[6])
    inputs = [tf.placeholder(tf.float32, [None, vocab_size]) for _ in xrange(num_steps)]
    outputs = {}
    hs = {}
    hs[-1] = tf.placeholder(tf.float32, [None, hidden_size])
    for i in range(len(inputs)):
        hs[i]      = tf.nn.tanh(tf.matmul(inputs[i], Wxh) + tf.matmul(hs[i-1], Whh) + bh)
        outputs[i] = tf.nn.softmax(tf.matmul(hs[i], Why) + by)

    # Implement Cross Entropy (Currently In[8] and In[9]):
    truths = [tf.placeholder(tf.float32, [None, vocab_size]) for _ in xrange(num_steps)]
    cross_entropies = [-tf.reduce_sum(truths[i]*tf.log(outputs[i])) for i in xrange(num_steps)]  # These operations act element-wise.
    cross_entropy = tf.add_n(cross_entropies)
    train_step = tf.train.AdagradOptimizer(learning_rate).minimize(cross_entropy)

    # Set up feed_dict
    feed_dict = ...

    # Run the batch (taken from RunModel in In[13]:
    return sess.run((train_step, hs[num_steps-1], cross_entropy), feed_dict=feed_dict)
```

Then, in `RunModel`, replace the `sess.run(...)` line with a call to the above function:

```
_, h_state, loss_out = RunBatch(input_chars, target_chars)
```

---

Please do let me know why you would want to do this; I'm still learning, too! 🙃

**Comment by rajarsheem** [comment]
**Why I am tending to use your custom rnn instead of the built in LSTM** ?
I am trying to design the LSTM (which is a just a not-so-complex extension of RNN) myself. So manually working in the loop where you have defined the recurrence gives some flexibility as you can explicitly monitor what's going on there.

**Why am I trying to use variable sequence ?**
Sequences can be of different length. (eg. seq2seq) When I am doing some other tasks other than char-rnn (seq2seq, classification), at each iteration we need to give a full sentence to the input which will differ in length every time. From more technical viewpoint, num_steps in
`inputs = [tf.placeholder(tf.float32, [None, vocab_size]) for _ in xrange(num_steps)]`
must be different for each iteration.  (It looks like the placeholders - inputs and truths should be inside the epoch loop somehow. Will this affect the session or graph somehow ?)

Looking at your code, it seems that you have caught my idea well. 
One thing I wanted you to ask (which is separate from the issue) is that if I change the recurrence to a LSTM (by adding some gates) will it still be different (or less efficient) from the built in LSTM ?
Secondly, will keeping the placeholders inside the epoch loop and creating them at each iteration will, anyhow, affect the learning or the tensorflow graph ?

**Comment by rajarsheem** [comment]
**And what I am particularly trying to do ?**
-> I want to extend your code to a LSTM and then pass variable length sentences to the input placeholder and after training will collect the last hidden state for each sentence (which is essentially an embedding for each sentence). Will see if I can visualise it.

Eventually, my aim is to construct a seq2seq.

**Comment by NHDaly** [comment]
> **Why I am tending to use your custom rnn instead of the built in LSTM?**
> I am trying to design the LSTM myself.

Oooh okay gotcha. Cool! I had thought about doing that, too, but I haven't read too much about LSTMs, and also I haven't worked on these side exploration projects in a while (I just moved to Chicago so I've been pretty busy. 😄 ) Sounds fun! Good luck; do post back if you have anything to share -- I'd be interested to see it!

> **Why am I trying to use variable sequence ?**
> ...  (eg. seq2seq) ... we need to give a full sentence to the input which will differ in length every time.

Woah, cool! I hadn't read about seq2seq before. It's really cool!

I had imagined you could do something similar by including an `end-of-sentence` marker. But even with something like that, I guess it's not necessary to have a variable `num_steps`. You could stop in the middle of a sentence, apply your backpropogation, and then continue parsing the sentence. But if you want to then pass it to a _new_ RNN, the _decoder_, it seems that yes, you need variable length sequences. Cool, this makes sense; thanks!

> Eventually, my aim is to construct a seq2seq.

!! :) Good luck, this seems really cool!

---

Now, to answer your questions:

> Looking at your code, it seems that you have caught my idea well.

Yes, I think that the change I suggested above is sufficient to allow variable-length sequences. This should work correctly! One thing you might want to play around with is whether or not to zero-out the `hidden_state` before each new training sequence. I don't do that currently, but it might be worthwhile since you're intending to represent each new sequence as a clean learning item.

> if I change the recurrence to a LSTM (by adding some gates) will it still be different (or less efficient) from the built in LSTM ?

I don't know. I have not done much digging into the TensorFlow implementation for LSTM. I believe it should be possible to recreate what they do. Will it be less efficient? Probably. I would guess that they implement large parts of it in C++ as one unit so as to be more efficient, but that's just a guess. Will it be different? Probably. Just because you aren't them, it's very likely your code will behave slightly differently.

Will it still work? Yeah! You should totally be able to implement your own LSTM! :)

> I want to extend your code to a LSTM and then pass variable length sentences to the input placeholder and after training will collect the last hidden state for each sentence (which is essentially an embedding for each sentence). Will see if I can visualise it.

Yeah, this sounds awesome!

**A recommendation:** I would be sure to only change one thing at a time while you're experimenting with this. It's always important to do that for Software Engineering, but it seems to be even more important for data science like this, where everything is so finicky.

First, I would leave it with my RNN implementation, and just try adding variable-length sequences. Then, if you get that to work, just try fixed-length sequences and your own custom LSTM. Then, if you get both those things working, try them together! :) Sounds fun!! Good luck!

**Comment by NHDaly** [comment]
~ Ah okay actually, you were right. The code solution I presented above actually won't work. The reason is the Adagrad update:

Adagrad update works by maintaining variables which more or less remember how many times the training step has been run so that it can degrade over time. TensorFlow implements this by creating Variables associated with any operation you are trying to minimize for your gradient descent, meaning the Adagrad Variables are associated with the `train_step`. So if you create a _new_ `train_step` when you resize the model (as we do above), you're creating new variables.

This is bad for two reasons: 1. it means adagrad isn't working (it's not remembering how many times it's been run). 2: TensorFlow requires you to initialize these new variables before you use them, so the above won't work.

I have thoughts on what's best, I'll post them in a bit.

But some possible solutions I see: 
1. try to avoid recreating the Adagrad update training step.
2. just don't use Adagrad, and switch to a more basic gradient descent, which doesn't decay over time.

**Comment by NHDaly** [comment]
One good idea, I think, for implementing #1 above, is to simply _not_ resize the model, but instead simulate it by having a model which you know is unrolled longer than any of your input sequences, and always just using a subsection of it.

For example, if you know your longest sequence is around 50 chars, you can start by creating a model unrolled to 100 steps just to be sure. Then, for example, for a sequence of length 13 (`Hello World.<EOS>`), you pad the remaining characters after the end-of-sequence character with garbage. This way, when you do backpropagation, the predictions after <EOS> will not contribute to the loss at all, and therefore, I think, will not contribute to the gradient. So it would be as if you trained only up to the <EOS> character.

For example,

 `Hello World.<EOS>\0\0\0\0\0\0\0\0\0\0\0...`

Now, instead of training all at once, you run in two steps. First, you run just up to generating the outputs (much like in hallucination) to get the output from the above sequence:

```
feed_dict={inputs[i]: input_chars[i] for i in range(len(input_chars))}
feed_dict.update({hs[-1]:h_state})
outs, final_h_state = sess.run((outputs,), feed_dict=feed_dict)
```

Maybe it might look like this: 

`elLooworLd.?mseesfo2adfav....`

Note that after the first 13 input characters, you just gave it `\0`, so it didn't know what to do; it just generates garbage. Now, you simulate the variable-sequence-length, by correcting the rest of the output characters after `<EOS>`:

```
outs = outs[:sequence_length-1] + '\0'*(100-sequence_length)
```

Now, `outs` will be "correct" after `<EOS>`: `elLooworLd.?\0\0\0\0\0\0\0...`

Finally, you finish the process by running the training step, but specifying the (modified) `outputs` tensor instead of the `inputs` this time.

```
feed_dict.update({truths[i]: target_chars[i] for i in range(len(target_chars))})
_, loss_out = sess.run((train_step, cross_entropy), feed_dict=feed_dict)
```

I think this will allow you to simulate having variable length inputs.

Oh, and if you want to get your "last" hidden state, you just actually grab `hs[seq_length-1]`. You should add that to the first `sess.run` statement, giving you:

```
outs, final_h_state = sess.run((outputs, hs[seq_length-1]), feed_dict=feed_dict)
```

---

The main downside to this method is it will be a bitter slower than necessary, because you are still doing the predictions for the fully unrolled model, even though you don't need it all. :/

**Comment by rajarsheem** [comment]
Your last suggestion is good though it makes me think something much better is still there hiding somewhere! 

One thing I would like to ask: In the original code,  you are taking `num_steps` placeholders in a list each having length `vocab_size`. Instead of this, can't you take a single placeholder of shape = `[None, vocab_size]` in which case you can freely feed unknown no. of word vectors each having size of `vocab_size` ?

I guess, in that case, you have to somehow iterate over the placeholder to compute the states and outputs.

**Comment by rajarsheem** [comment]
Btw, you can view my [attempt](https://github.com/rajarsheem/playing-with-RNNs/blob/master/rnn.ipynb). I think it can handle variable length-ed sequences. Although, I am not sure about the full correctness yet, but still I think it is working. Drop by suggestions!

**Comment by NHDaly** [comment]
Hey, sorry I'd never responded to your last comments! Hope you found success!!

I don't know if you saw this already, but it looks like TensorFlow released a new product specifically for addressing variable length input sequences!:
https://research.googleblog.com/2017/02/announcing-tensorflow-fold-deep.html

Cheers!

**Comment by rajarsheem** [comment]
No worries. I am long way past this ! 

Regards.
